Generate Client using OpenAPI spec tool, Update Setup Guide and Quickstart (#2)
ballerina-platform/module-ballerinax-hubspot.crm.lists: ballerina/Package.md

Reviewer: Do the changes suggested on `Module.md` file in here as well.

Author: Done. [a051732](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/a051732b30bd6d54faaed900cd46b0a4efa30bd4)

## Overview

[HubSpot](https://www.hubspot.com/) is an AI-powered customer relationship management (CRM) platform. 

> The ballerinax/hubspot.crm.lists offers APIs to connect and interact with the [HubSpot CRM Lists API](https://developers.hubspot.com/docs/reference/api/crm/lists) endpoints, specifically based on the [HubSpot CRM Lists API v3 OpenAPI specification](https://github.com/HubSpot/HubSpot-public-api-spec-collection/blob/main/PublicApiSpecs/CRM/Lists/Rollouts/144891/v3/lists.json)

Reviewer: ```suggestion
The `ballerinax/hubspot.crm.lists` offers APIs to connect and interact with the [HubSpot CRM Lists API](https://developers.hubspot.com/docs/reference/api/crm/lists) endpoints, specifically based on the [HubSpot REST API](https://developers.hubspot.com/docs/reference/api/overview)
```

## Setup guide

To use the HubSpot CRM Lists connector, you must have access to the HubSpot API through a HubSpot developer account and a HubSpot App under it. Therefore you need to register for a developer account at HubSpot if you don't have one already.

### Step 1: Create/Login to a HubSpot Developer Account

If you have an account already, go to the [HubSpot developer portal](https://app.hubspot.com/)

If you don't have a HubSpot Developer Account you can sign up to a free account [here](https://developers.hubspot.com/get-started)

> ### Step 2 (Optional): Create a [Developer Test Account](https://developers.hubspot.com/beta-docs/getting-started/account-types#developer-test-accounts) under your account
> 
> Within app developer accounts, you can create developer test accounts to test apps and integrations without affecting any real HubSpot data.
> 
> **_These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts._**

Reviewer: ```suggestion
### Step 2 (Optional): Create a Developer Test Account

Within app developer accounts, you can create a [developer test account](https://developers.hubspot.com/beta-docs/getting-started/account-types#developer-test-accounts) under your account to test apps and integrations without affecting any real HubSpot data.

> **Note:** These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts.
```

1. Go to Test Account section from the left sidebar.

>    <img src=../docs/setup/resources/test_acc_1.png alt="Hubspot developer portal" style="width: 70%;">

Reviewer: Please use the markdown format here. Check similar places.

```suggestion
![Hubspot developer portal](./docs/setup/resources/test_acc_1.png)
```

Author: Done. [3290d95](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/3290d952980ad9d39a0ebde71451862de5deea99)

2. Click Create developer test account.

   <img src=../docs/setup/resources/test_acc_2.png alt="Creating a hubspot test account" style="width: 70%;">

3. In the dialogue box, give a name to your test account and click create.

   <img src=../docs/setup/resources/test_acc_3.png alt="Adding a name and description for the test account" style="width: 70%;">

### Step 3: Create a HubSpot App under your account.

1. In your developer account, navigate to the "Apps" section. Click on "Create App"

   <img src=../docs/setup/resources/create_app_1.png alt="Creating a new App" style="width: 70%;">

2. Provide the necessary details, including the app name and description.

### Step 4: Configure the Authentication Flow.

1. Move to the Auth Tab.

   <img src=../docs/setup/resources/create_app_2.png alt="Selecting the Auth tab" style="width: 70%;">

2. In the Scopes section, add the following scopes for your app using the "Add new scope" button.

> 
>    `crm.lists.read`
> 
>    `crm.lists.write`
> 
>    `cms.membership.access_groups.write`

Reviewer: Use a list here.

```suggestion
    * `crm.lists.read`
    * `crm.lists.write`
    * `cms.membership.access_groups.write`
```

Author: Done. [199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

   <img src=../docs/setup/resources/scope_set.png alt="List of scopes" style="width: 70%;">

>    Scopes listed above are the mandotory scopes needed to use the HubSpot CRM Lists API. However you may need to add additional scopes based on your usecase. For example, if you are working with contacts, you may need to add `crm.objects.contacts.read` and `crm.objects.contacts.write` scopes as well.

Reviewer: We can remove this part.

````suggestion
````

Author: Done. [199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

Author: Had to add this part again because working with some endpoints needs other consent for other scopes as well. [2e03824](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/2e03824283846e1a70eca2beed35304f71548554)

> 3. Add your Redirect URI in the relevant section. You can also use localhost addresses for local development purposes. Click Create App.

Reviewer: ```suggestion
3. Add your Redirect URI in the relevant section. You can also use `localhost` addresses for local development purposes. Click "Create App".
```

Author: Done. [199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

   <img src=../docs/setup/resources/create_app_final.png alt="Adding the redirect URI and save" style="width: 70%;">

### Step 5: Get your Client ID and Client Secret

- Navigate to the Auth section of your app. Make sure to save the provided Client ID and Client Secret.

   <img src=../docs/setup/resources/get_credentials.png alt="Client ID and secret of the App" style="width: 70%;">

### Step 6: Setup Authentication Flow

Before proceeding with the Quickstart, ensure you have obtained the Access Token using the following steps:

1. Create an authorization URL using the following format:

   ```
   https://app.hubspot.com/oauth/authorize?client_id=<YOUR_CLIENT_ID>&scope=<YOUR_SCOPES>&redirect_uri=<YOUR_REDIRECT_URI>
   ```

   Replace the `<YOUR_CLIENT_ID>`, `<YOUR_REDIRECT_URI>` and `<YOUR_SCOPES>` with your specific value.

>     **_NOTE: If you are using a localhost redirect url, make sure to have a listner running at the relevant port before executing the next step. You can use [this gist](https://gist.github.com/lnash94/0af47bfcb7cc1e3d59e06364b3c86b59) and run it using `bal run`. Alternatively, you can use any other method to bind a listner to the port._**

Reviewer: Let's remove this part.
```suggestion
```

Author: Done. [199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

2. Paste it in the browser and select your developer test account to intall the app when prompted. Provide consent for all scopes needed.

   <img src=../docs/setup/resources/install_app.png alt="Selecting the App for OAuth process" style="width: 70%;">

3. A code will be displayed in the browser. Copy the code.

> 
>    ```
>    Received code: na1-129d-860c-xxxx-xxxx-xxxxxxxxxxxx
>    ```

Reviewer: Remove this part.

```suggestion
```

Author: Done.[199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

     "token_type": "bearer",
     "refresh_token": "<Refresh Token>",
     "access_token": "<Access Token>",
     "expires_in": 1800
   }
   ```

5. Store the access token securely for use in your application.

## Quickstart

To use the `HubSpot CRM Lists` connector in your Ballerina application, update the `.bal` file as follows:

### Step 1: Import the module

Import the `hubspot.crm.lists` module and `oauth2` module.

```ballerina
import ballerinax/hubspot.crm.lists as crmlists;
import ballerina/oauth2;
```

### Step 2: Instantiate a new connector

1. Create a `Config.toml` file and, configure the obtained credentials in the above steps as follows:

   ```toml
    clientId = <Client Id>
    clientSecret = <Client Secret>
    refreshToken = <Refresh Token>
   ```

2. Instantiate a `OAuth2RefreshTokenGrantConfig` with the obtained credentials and initialize the connector with it.

    ```ballerina 
    configurable string clientId = ?;
    configurable string clientSecret = ?;
    configurable string refreshToken = ?;

    OAuth2RefreshTokenGrantConfig auth = {
        clientId,
        clientSecret,
        refreshToken,
        credentialBearer: oauth2:POST_BODY_BEARER
    };

>     final string serviceUrl = "https://api.hubapi.com";
> 
>     final crmlists:Client crmListClient = check new Client(config = {auth}, serviceUrl = serviceUrl);

Reviewer: We can omit the service URL here.

```suggestion
    final crmlists:Client crmListClient = check new Client(config = {auth});
```

Author: Done. [199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)
